eras/shelley/impl/src/Cardano/Ledger/Shelley/LedgerState.hs

```diff
@@ -1153,45 +1158,89 @@ startStep slotsPerEpoch b@(BlocksMade b') es@(EpochState acnt ss ls pr _ nm) max
       Coin rPot = _feeSS ss <> deltaR1
       deltaT1 = floor $ unboundRational (getField @"_tau" pr) * fromIntegral rPot
       _R = Coin $ rPot - deltaT1
+
+      -- We now compute stake pool specific values that are needed for computing
+      -- member and leader rewards.
+      activestake = fold . unStake $ stake'
       totalStake = circulation es maxSupply
+      mkPoolRewardInfoCurry =
+        mkPoolRewardInfo
+          (getField @"_d" pr, getField @"_a0" pr, getField @"_nOpt" pr)
+          _R
+          b
+          (fromIntegral blocksMade)
+          stake'
+          delegs'
+          totalStake
+          activestake
+
+      -- We map over the registered stake pools to compute the revelant
+      -- stake pool specific values.
+      allPoolInfo = Map.map mkPoolRewardInfoCurry poolParams
+
+      -- Stake pools that do not produce any blocks get no rewards,
+      -- but some information is still needed from non-block-producing
+      -- pools for the ranking algorithm used by the wallets.
+      blockProducingPoolInfo = Map.mapMaybe rightToMaybe allPoolInfo
+
+      getSigma = unStakeShare . poolRelativeStake
+      makeLikelihoods = \case
+        -- This pool produced no blocks this epoch
+        Left (StakeShare sigma) ->
+          likelihood
+            0
+            (leaderProbability asc sigma $ getField @"_d" pr)
+            slotsPerEpoch
+        -- This pool produced at least one block this epoch
+        Right info ->
+          likelihood
+            (poolBlocks info)
+            (leaderProbability asc (getSigma info) $ getField @"_d" pr)
+            slotsPerEpoch
+      newLikelihoods = Map.map makeLikelihoods allPoolInfo
+
+      -- We now compute the leader rewards for each stake pool.
+      genRewardAcnt = getRwdCred . _poolRAcnt . poolPs
+      collectLRs acc poolRI =
+        if genRewardAcnt poolRI `Map.member` _rewards ds
+          then
+            Map.insertWith
+              Set.union
```

Comment:
@nc6 it is not an insert into a singleton set, but rather an insertion of a singleton set into a map that can have a set present, in which case it will be unioned.

This is the same as:

```haskell
  let pr = poolLeaderReward poolRI
  in Map.alter (Just . maybe (Set.singleton pr) (Set.insert pr)) (genRewardAcnt poolRI) acc
```

Reply:
Yeah, for some reason it took me a long time to understand this function!